File: tools/skill-loader-node/src/loader.js
import fs from 'node:fs/promises';
import path from 'node:path';
import YAML from 'yaml';

async function fileExists(p) {
  try {
    await fs.access(p);
    return true;
  } catch {
    return false;
  }
}

export async function discoverManifestFile(skillDir) {
  const candidates = ['SKILL.md', 'skill.json', 'skill.yaml', 'skill.yml', 'package.json'];
  for (const name of candidates) {
    const full = path.join(skillDir, name);
    if (!(await fileExists(full))) continue;

    if (name === 'package.json') {
      try {
        const raw = await fs.readFile(full, 'utf8');
        const pkg = JSON.parse(raw);
        if (pkg.skill) return full;
        continue;
      } catch {
        continue;
      }
    }

    return full;
  }
  return null;
}

export async function loadManifest(skillDir) {
  const file = await discoverManifestFile(skillDir);
  if (!file) return null;

  const base = path.basename(file);
  const raw = await fs.readFile(file, 'utf8');

  if (base === 'SKILL.md') {
    const match = raw.match(/^---\s*\n([\s\S]*?)\n---\s*\n/);
    const manifest = match?.[1] ? YAML.parse(match[1]) : {};
    return { manifest, source: 'SKILL.md' };
  }

  if (base === 'skill.json') {
    const manifest = JSON.parse(raw);
    return { manifest, source: 'skill.json' };
  }

  if (base === 'skill.yaml' || base === 'skill.yml') {
    const manifest = YAML.parse(raw);
    return { manifest, source: base };
  }

  if (base === 'package.json') {
    const pkg = JSON.parse(raw);
    if (!pkg.skill) return null;
    if (typeof pkg.skill !== 'object')
      throw new Error('package.json skill field must be an object');
    return { manifest: pkg.skill, source: 'package.json#skill' };
  }

  throw new Error(`Unsupported manifest file: ${file}`);
}

export function validateManifest(manifest) {
  const missing = [];
  if (manifest?.schemaVersion || manifest?.id) {
    for (const k of ['id', 'name']) {
      if (!manifest?.[k]) missing.push(k);
    }
  }
  if (missing.length) throw new Error(`Manifest missing required fields: ${missing.join(', ')}`);
}

export function buildRegistrationInfo(manifest, skillDir, { runtime = 'node' } = {}) {
  const entryPath =
    runtime === 'node'
      ? (manifest?.entry?.node ?? manifest?.entry?.path)
      : runtime === 'python'
        ? (manifest?.entry?.python ?? manifest?.entry?.path)
        : manifest?.entry?.path;

  let id = manifest.id;
  if (!id) {
    const absPath = path.resolve(skillDir);
    const parts = absPath.split(path.sep);
    const extIdx = parts.lastIndexOf('external_skills');
    if (extIdx !== -1 && parts.length > extIdx + 1) {
      const repoDir = parts[extIdx + 1];
      if (repoDir.includes('__')) {
        const [owner, repo] = repoDir.split('__');
        const subpath = parts.slice(extIdx + 2).join('::');
        id = subpath ? `${owner}::${repo}::${subpath}` : `${owner}::${repo}`;
      }
    }
  }

  return {
    id: id ?? `local::${path.basename(skillDir)}`,
    name: manifest.name ?? path.basename(skillDir),
    version: manifest.version ?? '0.0.0',
    description: manifest.description ?? null,
    dir: path.resolve(skillDir),
    entry: {
      runtime,
      path: entryPath ?? null,
      exports: manifest?.entry?.exports ?? {},
    },
    repository: manifest.repository ?? null,
    integrity: manifest.integrity ?? null,
  };
}

export function createRegistry() {
  const byId = new Map();
  return {
    register(info) {
      if (!info?.id) throw new Error('registry.register requires info.id');
      byId.set(info.id, info);
      return info;
    },
    get(id) {
      return byId.get(id) ?? null;
    },
    list() {
      return Array.from(byId.values());
    },
  };
}

async function walkAndDiscover(skillDir, results = []) {
  const manifest = await discoverManifestFile(skillDir);
  if (manifest) {
    results.push(skillDir);
  }

  let entries = [];
  try {
    entries = await fs.readdir(skillDir, { withFileTypes: true });
  } catch (e) {
    return results;
  }

  for (const entry of entries) {
    if (entry.isDirectory()) {
      const name = entry.name;
      if (name === '.git' || name === 'node_modules' || name === 'dist' || name === 'build')
        continue;

      const isAllowedHiddenDir = ['.claude', '.agents', '.config', '.github'].includes(name);
      if (name.startsWith('.') && !isAllowedHiddenDir) {
        continue;
      }

      await walkAndDiscover(path.join(skillDir, name), results);
    }
  }
  return results;
}

export async function loadAndRegister({ skillDir, registry, runtime = 'node' }) {
  const skillDirs = await walkAndDiscover(skillDir);

  const results = [];
  for (const dir of skillDirs) {
    try {
      const loaded = await loadManifest(dir);
      if (!loaded) continue;
      const { manifest, source } = loaded;
      validateManifest(manifest);
      const info = buildRegistrationInfo(manifest, dir, { runtime });
      info.manifestSource = source;
      results.push(registry.register(info));
    } catch (e) {
      console.error(`Failed to load skill in ${dir}: ${e.message}`);
    }
  }

  if (results.length === 0) {
    throw new Error(`No valid manifest found in ${skillDir} or its subdirectories`);
  }

  return results.length === 1 ? results[0] : results;
}

File: tools/skill-loader-py/loader.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

import yaml


def discover_manifest_file(skill_dir: str) -> Path | None:
    base = Path(skill_dir)
    for name in ("SKILL.md", "skill.json", "skill.yaml", "skill.yml", "package.json"):
        p = base / name
        if p.exists():
            return p
    return None


def load_manifest(skill_dir: str) -> tuple[dict, str]:
    p = discover_manifest_file(skill_dir)
    if not p:
        raise FileNotFoundError(f"No manifest found in {skill_dir}")

    raw = p.read_text(encoding="utf-8")
    if p.name == "SKILL.md":
        import re
        m = re.match(r"^---\s*\n(.*?)\n---\s*\n", raw, flags=re.S)
        if m:
            return yaml.safe_load(m.group(1)) or {}, "SKILL.md"
        return {}, "SKILL.md"
    if p.name == "skill.json":
        return json.loads(raw), "skill.json"
    if p.name in ("skill.yaml", "skill.yml"):
        return yaml.safe_load(raw), p.name
    if p.name == "package.json":
        pkg = json.loads(raw)
        if "skill" not in pkg:
            raise ValueError('package.json found but missing top-level "skill" field')
        if not isinstance(pkg["skill"], dict):
            raise ValueError("package.json skill field must be an object")
        return pkg["skill"], "package.json#skill"

    raise ValueError(f"Unsupported manifest file: {p}")


def validate_manifest(manifest: dict) -> None:
    # SKILL.md based skills might not have schemaVersion or id or version or entry
    # They are more like reference documentation than executable packages
    if "name" not in manifest and "id" not in manifest:
         # Minimal validation: must have at least a name or we use folder name later
         pass


def build_registration_info(manifest: dict, skill_dir: str, *, runtime: str = "python") -> dict:
    entry = manifest.get("entry") or {}
    if runtime == "node":
        entry_path = entry.get("node") or entry.get("path")
    elif runtime == "python":
        entry_path = entry.get("python") or entry.get("path")
    else:
        entry_path = entry.get("path")

    return {
        "id": manifest.get("id"),
        "name": manifest.get("name"),
        "version": manifest.get("version"),
        "description": manifest.get("description"),
        "dir": str(Path(skill_dir).resolve()),
        "entry": {"runtime": runtime, "path": entry_path, "exports": entry.get("exports") or {}},
        "repository": manifest.get("repository"),
        "integrity": manifest.get("integrity"),
    }


@dataclass
class Registry:
    _by_id: dict

    @classmethod
    def create(cls) -> "Registry":
        return cls(_by_id={})

    def register(self, info: dict) -> dict:
        if not info.get("id"):
            raise ValueError("registry.register requires info.id")
        self._by_id[info["id"]] = info
        return info

    def get(self, skill_id: str) -> dict | None:
        return self._by_id.get(skill_id)

    def list(self) -> list[dict]:
        return list(self._by_id.values())


def load_and_register(*, skill_dir: str, registry: Registry, runtime: str = "python") -> dict:
    manifest, source = load_manifest(skill_dir)
    validate_manifest(manifest)
    info = build_registration_info(manifest, skill_dir, runtime=runtime)
    info["manifestSource"] = source
    return registry.register(info)
